### PR TITLE
[Spree Upgrade] Add migrations necessary to successfully migrate data to v2

### LINKED ISSUE #3600: [Spree Upgrade] Fix shipments.stock_location_id migration issue
## Description
After migrating to v2, shipment.stock_location_id become null.

## Expected Behavior
After migrating v1 data to v2, we dont see any shipment.stock_location_id nulls.

## Actual Behaviour
shipment.stock_location_id is null for all shipments after migration.

## Steps to Reproduce
Fetch live data in v1 model, run db:migrate, inspect db.

## Possible Fix
Because we only have one stock location, we can make all existing shipments get the default stock location.
To do that we need to bring spree/spree@219e09b from spree v2.1.0

What's the best way to include it? Shall we simply copy the migration code to our branch as a new migration in OFN?

### PR DESCRIPTION
**Pull request by luisramos0**
#### What? Why?

Closes #3600 and #3594 

Add migrations necessary to successfully migrate live data to v2.

#### What should we test?
Migrating live data to v2 should keep the app working, namely:
- all shipments should have stock location defined
- all shipping methods and products should have a shipping category

#### Dependencies
This is the continuation of #3591

### CODE REVIEW

File: db/migrate/20190313142051_set_default_stock_location_on_shipments.rb [hunk 1]
```diff
@@ -0,0 +1,8 @@
+class SetDefaultStockLocationOnShipments < ActiveRecord::Migration
+  def up
+    if Spree::Shipment.where('stock_location_id IS NULL').count > 0
+      location = DefaultStockLocation.find_or_create
```
Comment (mkllnk) [inline db/migrate/20190313142051_set_default_stock_location_on_shipments.rb]: I would have gone the other way, defining `StockLocation` and `Shipment` here to be independent of code changes. But I'm happy to go with this for now.

File: db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb [hunk 2]
```diff
@@ -0,0 +1,13 @@
+class AddShippingCategoryToShippingMethodsAndProducts < ActiveRecord::Migration
+  def up
+    # This is different from the equivalent Spree migration
+    #   Here we are creating the default shipping category even if there are already shipping categories
+    default_category = Spree::ShippingCategory.create!(name: "Default")
```
Comment (mkllnk) [inline db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: Is it safe to assume that we don't have a shipping category yet? I guess so.

File: db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+    Spree::ShippingMethod.all.each do |method|
```
Comment (mkllnk) [inline db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: I know this comes from Spree and changing it is out of scope. But just for similar cases in the future `Spree::ShippingMethod.find_each` would be better as it iterates in batches when there are lots of records. https://api.rubyonrails.org/classes/ActiveRecord/Batches.html#method-i-find_each

It only matters if there are more than 1000 records and we have ~500 in Australia. So it should be fine.

File: db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb (hunk 2 above)
Comment (luisramos0) [reply db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: Even if we have other shipping categories (we will in some production environments), I think it's good that we get this default one. We can then use it to make shipping category optional at some point...

File: db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb (hunk 3 above)
Comment (luisramos0) [reply db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: nice, thanks for pointing that out!

File: db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb (hunk 2 above)
Comment (mkllnk) [reply db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: Okay, let me rephrase the question: Is it safe to assume that we don't have a `Default` shipping category yet? I guess this would fail then.

File: db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb (hunk 3 above)
Comment (sauloperez) [reply db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: that applies to all updates we perform. I checked the one SetDefaultStockLocationOnShipments and I don't think it'll be a problem for the numbers we have so I prefer to lean towards code simplicity instead of performance. Agree on this?

File: db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb (hunk 2 above)
Comment (sauloperez) [reply db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: There is neither an ActiveModel validation nor a DB constraint so we will have duplicates at most. Even in that case, I think we could just merge them later on right?

Comment (luisramos0) [reply db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: ah, I see you question now Maikel.
Default duplicates will not be a problem, all we need here is a shipping category to assign to all products and methods without a shipping category. after this, the users will have to set shipping categories on products and methods as they create them. and admins will be able to manage shipping categories as they want (including this Default one).

Comment (luisramos0) [reply db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: I agree with Pau, this will never fail as there is no constraint on unique shipping_category.name. And if we end up with duplicate names (there's no "Default" shipping category in any of our main instances right now), admins will be able to manage them as they want.

File: db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb (hunk 3 above)
Comment (luisramos0) [reply db/migrate/20190313142540_add_shipping_category_to_shipping_methods_and_products.rb]: :+1:

**Comment by luisramos0** [comment]
you guys have approved the PR so I assume you are ok with the two points raised above. I have tested this with AUS data, it works. So this is ready to be merged. Correct me if I am wrong.
